File: scripts/git_automator.js
import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';

function runCmd(cmd, check = true) {
  console.log(`Running: ${cmd}`);
  try {
    const output = execSync(cmd, { encoding: 'utf-8', stdio: 'pipe' });
    return { stdout: output, status: 0 };
  } catch (error) {
    if (check) {
      console.error(`Error executing command: ${cmd}`);
      console.error(error.stderr || error.message);
      process.exit(1);
    }
    return { stdout: error.stdout, status: error.status || 1 };
  }
}

function safeWrite(filePath, content) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(filePath, content, 'utf-8');
}

async function processTasks(jsonPath, localOnly = false) {
  const tasks = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));

  for (let i = 0; i < tasks.length; i++) {
    const task = tasks[i];
    console.log(`\n--- Processing Task ${i + 1}/${tasks.length}: ${task.branch} ---`);

    // Checkout main
    runCmd('git checkout main');

    // Create/Switch Branch
    const branchCheck = runCmd(`git rev-parse --verify ${task.branch}`, false);
    if (branchCheck.status === 0) {
      runCmd(`git checkout ${task.branch}`);
    } else {
      runCmd(`git checkout -b ${task.branch}`);
    }

    // Apply Changes
    if (task.changes) {
      for (const change of task.changes) {
        safeWrite(change.path, change.content);
      }
    }

    // Commit
    runCmd('git add .');
    const statusResult = runCmd('git status --porcelain');
    if (!statusResult.stdout.trim()) {
      console.log('No changes to commit. Skipping.');
    } else {
      // Use a temp file for commit message to avoid shell escaping issues
      fs.writeFileSync('.commit_msg.tmp', task.commit_msg);
      runCmd(`git commit -F .commit_msg.tmp`);
      fs.unlinkSync('.commit_msg.tmp');
    }

    // PR Flow
    if (!localOnly) {
      try {
        runCmd(`git push origin ${task.branch}`);
        const prTitle = task.commit_msg;
        const prBody = task.pr_body || "Automated micro-commit.";
        
        fs.writeFileSync('.pr_body.tmp', prBody);
        runCmd(`gh pr create --title "${prTitle}" --body-file .pr_body.tmp --base main --head ${task.branch}`, false);
        fs.unlinkSync('.pr_body.tmp');
        
        runCmd(`gh pr merge ${task.branch} --merge --auto --delete-branch`, false);
      } catch (e) {
        console.error(`Remote operation failed for ${task.branch}: ${e.message}`);
        // Fallback to local merge
        runCmd('git checkout main');
        runCmd(`git merge ${task.branch} --no-ff -m "Merge ${task.branch}"`);
      }
    } else {
      runCmd('git checkout main');
      runCmd(`git merge ${task.branch} --no-ff -m "Merge ${task.branch}"`);
    }

    console.log(`--- Completed Task ${i + 1} ---\n`);
  }
}

const args = process.argv.slice(2);
const jsonPath = args[0];
const localFlag = args.includes('--local');

if (!jsonPath) {
  console.error('Please provide a JSON task file path.');
  process.exit(1);
}

processTasks(jsonPath, localFlag);

File: scripts/git_automator.py
import os
import subprocess
import json
import time

def run_cmd(cmd, check=True):
    print(f"Running: {cmd}")
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    if result.returncode != 0 and check:
        print(f"Error: {result.stderr}")
        raise Exception(f"Command failed: {cmd}")
    return result

def safe_write(path, content):
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, 'w', encoding='utf-8') as f:
        f.write(content)

def process_tasks(json_file, local_only=False):
    with open(json_file, 'r') as f:
        tasks = json.load(f)

    for i, task in enumerate(tasks):
        print(f"--- Processing Task {i+1}/{len(tasks)}: {task.get('branch', 'unknown')} ---")
        branch = task['branch']
        changes = task.get('changes', [])
        commit_msg = task['commit_msg']
        pr_body = task.get('pr_body', "Automated micro-commit.")

        # Checkout main and update
        run_cmd("git checkout main")
        # if not local_only:
        #    run_cmd("git pull origin main") # Optional, to be safe

        # Create/Switch Branch
        # Check if branch exists
        res = run_cmd(f"git rev-parse --verify {branch}", check=False)
        if res.returncode == 0:
            run_cmd(f"git checkout {branch}")
        else:
            run_cmd(f"git checkout -b {branch}")

        # Apply Changes
        for change in changes:
            safe_write(change['path'], change['content'])
        
        # Commit
        run_cmd("git add .")
        # Check if anything to commit
        status = run_cmd("git status --porcelain")
        if not status.stdout.strip():
            print("No changes to commit. Skipping.")
        else:
            run_cmd(f'git commit -m "{commit_msg}"')

        # PR Flow
        if not local_only:
            try:
                run_cmd(f"git push origin {branch}")
                # Create PR
                # Check if PR already exists
                # This could fail if PR exists, so we ignore error or check first
                run_cmd(f'gh pr create --title "{commit_msg}" --body "{pr_body}" --base main --head {branch}', check=False)
                # Merge PR
                run_cmd(f"gh pr merge {branch} --merge --auto --delete-branch")
                # Wait a bit for GitHub to process?
                # time.sleep(2)
            except Exception as e:
                print(f"Remote operation failed: {e}")
                # Fallback to local merge to keep history moving
                run_cmd("git checkout main")
                run_cmd(f"git merge {branch}")
        else:
            # Local Merge
            run_cmd("git checkout main")
            run_cmd(f"git merge {branch} --no-ff -m \"Merge {branch}\"")

        print(f"--- Completed Task {i+1} ---\n")

if __name__ == "__main__":
    import sys
    mode = False
    if len(sys.argv) > 2 and sys.argv[2] == "--local":
        mode = True
    
    process_tasks(sys.argv[1], local_only=mode)
